### PR TITLE
feat: add ability to provide function in identity requests (fixes #177)

### LINKED ISSUE #177: Allow value in identity requests to be a function
Currently in identity requests it's possible to pass a `Promise` into `value`, which is awesome, but if promise is returned from function, this function is called every time route changes, which leads to extra requests, even if user id didn't change, e.g. when navigating from `/user/:id/pulls` to `/user/:id/commits`

```js
const fetchUser = id => fetch(`/users/${id}`);

connect(props => ({
  user: {
    value: fetchUser(props.match.params.id),
    comparison: props.match.params.id
  }
}))(Profile);
```

This can be solved if `value` can be function:

```js
const fetchUser = id => fetch(`/users/${id}`);

connect(props => ({
  user: {
    value: () => fetchUser(props.match.params.id),
    comparison: props.match.params.id
  }
}))(Profile);
```

Function passed into `value` should only be invoked if comparison changed.
What do you think?

### PR DESCRIPTION
**Pull request by ipanasenko**
This PR fixes #177

It's now possible to pass a function to identity request's `value` prop:
```js
const fetchUser = id => fetch(`/users/${id}`);

connect(props => ({
  user: {
    value: () => fetchUser(props.match.params.id),
    comparison: props.match.params.id
  }
}))(Profile);
```

### CODE REVIEW

**Comment by ryanbrainard** [comment]
Thanks for working on this! Looks good at first glance, but will look into the tests and corner cases a little more later.

Since someone could have theoretically provided a function before, I guess this should be considered a breaking change..

Oh, can you update the API doc and readme for this change too?

**Comment by ipanasenko** [comment]
@ryanbrainard updated docs

**Comment by slorber** [comment]
@ipanasenko @ryanbrainard can we publish this as beta on npm? I'd be happy to test that implementation on a real project

**Comment by ipanasenko** [comment]
@slorber I was very lazy, so I just built dist and copied it into my project's node_modules. You can fetch my branch and do the same ;)

**Comment by ipanasenko** [comment]
@slorber or modify your package.json to look like this: `"react-refetch": "ipanasenko/react-refetch#feat/identity-function-dist",`

**Comment by ryanbrainard** [comment]
@slorber @ipanasenko I just published this as a beta release: `react-refetch@2.0.0-0`

**Comment by slorber** [comment]
thanks,

i've just tested and it seems to work fine

File: test/components/connect.spec.js [hunk 1]
```diff
@@ -329,6 +329,135 @@ describe('React', () => {
...
+        const stub = TestUtils.findRenderedComponentWithType(container, Passthrough)
+        expect(stub.props.testFetch).toIncludeKeyValues({
+          fulfilled: true, pending: false, refreshing: false, reason: null, rejected: false, settled: true, value: 'foo', meta: { test: 'voodoo' }
+        })
+
+        done()
+      })
+    })
+
+    it('should invoke value() only if `comparison` changed', (done) => {
```
Comment (ryanbrainard) [inline test/components/connect.spec.js]: test name dupe. this looks like this was an accidential copy/paste of another test

File: README.md [hunk 2]
```diff
@@ -329,7 +329,17 @@ connect(props => ({
...
 
-**Identity requests can also be provided a `Promise` or thenable.** In this case, the `PromiseState` will be `pending` until the `Promise` is resolved. This can be helpful for asynchronous, non-fetch operations (e.g. file i/o) that want to use a similar pattern as fetch operations.
+**Identity requests can also be provided a `Promise`/thenable or a `Function`.**
+In case of a `Promise`, the `PromiseState` will be `pending` until the `Promise` is resolved. This can be helpful for asynchronous, non-fetch operations (e.g. file i/o) that want to use a similar pattern as fetch operations.
+In case of a `Function`, value returned from it will be used instead, as in cases described above. `Function` will be only be called when `comparison` changed.
+
+```jsx
+connect(props => ({
+  usersFetch: {
+    value: () => someAPI.list('/users')
```
Comment (ryanbrainard) [inline README.md]: What about changing to fetching a single user with a `comparison` on the `props.userId` so it's more obvious when the function is called again?

File: test/components/connect.spec.js [hunk 3]
```diff
@@ -329,6 +329,135 @@ describe('React', () => {
...
+      const renderSpy = expect.createSpy(() => ({}))
+      function render() {
+        renderSpy()
+        return <Passthrough/>
+      }
+
+      const valueSpy = expect.createSpy(() => ({}))
+      @connect(({ foo }) => ({
+        testFetch: {
+          value: valueSpy,
```
Comment (ryanbrainard) [inline test/components/connect.spec.js]: even though it won't actually change the behavior of the test, consider changing to `value: () => { valueSpy() }` to better test the function being created inside (like most apps would do), while still having the spy outside (so your test works)

**Comment by ryanbrainard** [comment]
@ipanasenko I just pushed up some changes to https://github.com/heroku/react-refetch/compare/stage-2.0.0. I updated the docs and tests as I mentioned in my review. If you're happy with them, we could merge them as-is (or of course push any changes you don't like). Then we can do the warnings (if we want them... interested to hear your thoughts) in a separate PR.

**Comment by slorber** [comment]
@ryanbrainard maybe it's worth issuing a warning in dev in case function change over time and comparison is not provided

**Comment by ryanbrainard** [comment]
I finally dusted this off and added the `'Request object with functional 'value' must also declare 'comparison'` invariant to the `stage-2.0.0` branch. I'll merge that into master shortly.

**Comment by ryanbrainard** [comment]
Published as beta v2.0.0-3

**Comment by ryanbrainard** [comment]
Published in v2.0.0
